Cross Compiles CI: Run evp tests on pull requests

t8m opened the pull request:
Also fixes a CI regression on ARM 32 bit on master branch from https://github.com/openssl/openssl/pull/21511

t8m commented:
The second patch is of course only for the master branch.

slontis commented on crypto/sha/asm/keccak1600-armv4.pl:
```diff
@@ -947,13 +947,13 @@ sub Round {
 .align	5
 SHA3_squeeze:
-	stmdb	sp!,{r0,r3-r11,lr} @ push 11 registers
+	stmdb	sp!,{r0,r3-r10,lr}
```
Do you know why it doesn't save r12? It trashes this with $bsz

t8m replied:
According to https://github.com/ARM-software/abi-aa/blob/main/aapcs32/aapcs32.rst:

_Register r12 (IP) may be used by a linker as a scratch register between a routine and any subroutine it calls (for details, see [Use of IP by the linker](https://github.com/ARM-software/abi-aa/blob/main/aapcs32/aapcs32.rst#use-of-ip-by-the-linker)). It can also be used within a routine to hold intermediate values between subroutine calls._

So it does not need to be saved.

slontis commented on crypto/sha/asm/keccak1600-armv4.pl:
```diff
@@ -933,7 +933,7 @@ sub Round {
 }
 
-{ my ($out,$len,$A_flat,$bsz,$next) = map("r$_", (4,5,10,12,11));
+{ my ($out,$len,$A_flat,$bsz,$next) = map("r$_", (4,5,10,12,0));
```
It seems reasonable to use r0. 
It is used down on line 984 as temp, so this should be ok.

slontis commented on crypto/sha/build.info:
```diff
@@ -169,15 +169,16 @@ GENERATE[keccak1600-s390x.S]=asm/keccak1600-s390x.pl
 
-# These are not yet used
+# These are not yet used and do not support multi-squeeze
+GENERATE[keccak1600-c64x.S]=asm/keccak1600-c64x.pl
```
Note there are also references to neon in the asm code.. I could not see how it got to these functions.

t8m replied:
Yeah, these are unused.

slontis commented on .github/workflows/cross-compiles.yml:
```diff
@@ -210,3 +210,9 @@ jobs:
+      if: github.event_name == 'pull_request' && matrix.platform.tests != 'none'
+      run: |
+        make test HARNESS_JOBS=${HARNESS_JOBS:-4} \
+                  TESTS="test_evp*" \
```
This at least does something.. It probably wouldnt have found the issue with my first 'fix'. but better than not doing anything.

t8m replied:
It would find it. The CI on this PR fails without the fix.

slontis commented on .github/workflows/cross-compiles.yml:
```diff
@@ -210,3 +210,9 @@ jobs:
                   TESTS="${{ matrix.platform.tests }} -test_afalg" \
                   QEMU_LD_PREFIX=/usr/${{ matrix.platform.arch }}
+    - name: make evp tests
+      if: github.event_name == 'pull_request' && matrix.platform.tests != 'none'
```
Should this also be ( || 'push") && (matrix....)

t8m replied:
No, in that case the evp tests are run as part of the more complete set.

beldmit reviewed (commented):
I understand and approve the CI part but don't understand the ASM part :)

t8m commented:
> I understand and approve the CI part but don't understand the ASM part :)

I changed the mapping of `$next` to `r0` thus `r11` does not need to be saved and fixed the bug loading of `$next` from wrong place on stack.

mattcaswell commented:
You suggest in this [comment](https://github.com/openssl/openssl/issues/22757#issuecomment-1818538532) that you will make an update to this PR...but I don't see any changes since you made that comment?

t8m commented:
> You suggest in this [comment](https://github.com/openssl/openssl/issues/22757#issuecomment-1818538532) that you will make an update to this PR...but I don't see any changes since you made that comment?

I am still investigating that. IMO it will be better if I open a new PR for these changes, if they are needed.

mattcaswell reviewed (approved):
My review of the assembler is "superficial"

slontis commented on crypto/sha/asm/keccak1600-ppc64.pl:
```diff
@@ -668,7 +668,7 @@
 	mr	$len,r5
 	mr	$bsz,r6
-	${UCMP}i r7,0                   ; r7 = 'next' argument
+	cmplwi	r7,0                    ; r7 = 'next' argument
```
Looks ok to me.. Took me a while to understand this instruction...

cmplwi = compare logical word immediate (does an unsigned compare of a 32 bit register with a 16 bit immediate value that is zero extended)
cmpldi = compare logical double word immediate
cmpwi = signed compare

(cmpldi rA, UI maps to cmpli cr0, 0, rA, u16)

t8m replied:
For PPC64 w = 32bit d = 64bit. And that is the intention - the next argument is int -> 32 bit.

slontis commented on crypto/sha/asm/keccak1600-x86_64.pl:
```diff
@@ -524,7 +524,7 @@
 	mov	%rdx,$len
 	mov	%rcx,$bsz
-	bt	\$0,$next
+	bt	\$0,${next}d
```
R8 to R15 (d = 32 bit, w = 16 bit) So look ok.

openssl-machine commented:
This pull request is ready to merge

hlandau commented:
Merged to master. Thank you.

It does not look like this merges cleanly to 3.0, 3.1 or 3.2. @t8m, do you want to investigate and raise another PR?

t8m commented:
Only the first commit was intended for the older branches. I've cherry-picked it to the 3.2, 3.1, and 3.0 branches.
